Add cbmc primitive regression tests

esteffin opened the pull request:
This PR is composed of 3 parts (commits):
 1. Run all now working `cbmc-primitive` regression tests with the incremental SMT2 solver,
 2. Fix an issue in the incremental SMT2 backed converter when a pointer offset was wrongly considered unsigned,
 3. Run other 3 now working (fixed by step 2) `cbmc-primitive` regression tests with the incremental SMT2 solver.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Fix pointer offset issue as converted as unsigned

tautschnig commented on regression/cbmc-primitives/dynamic-object-02/test-no-cp.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE new-smt-backend
```
Random place for comment: would it be possible to provide any indication which commit made things work?

esteffin replied:
It's hard as it was the combination of 2 or more PRs.
Namely the one adding support for `byte_(update|extract)_exprt` and the dynamic memory slices.

I will write in the commit body, but I cannot link a specific commit/PR as there are multiple.

tautschnig commented on regression/cbmc-primitives/pointer-offset-01/test-no-cp.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE new-smt-backend
```
Please fold updates to test into the commit that fixes the problem. Otherwise it is very hard to understand what effect a change has.

esteffin replied:
I can, however the third commit contains regression tests that need the second commit fix to work.
So I decided to keep 
 - Commit 1: add regression tests that were working because of previous PRs (see other comment)
 - Commit 2: fix issue with unsigned pointer offset
 - Commit 3: add regression tests now working because of commit 2

Alternatively I can move commit 2 at the beginning and squash commit 1 and 3

tautschnig replied:
My apologies, my comment was not as clear as should have been. What I'd ask for is squashing commits 2 and 3.

Commit:
Run working cbmc-primitive regressions with SMT2

Adding all cbmc-primitive regression tests that are now passing as the
necessary support has been added. Namely the regression tests using
dynamic memory, byte_extract_exprt and byte_update_exprt are now
supported and have been added.
